libshortfin/src/shortfin/local/scheduler.cc

```diff
@@ -0,0 +1,180 @@
+
+Account::Account(Device *device)
+    : device_(device), hal_device_(device->hal_device()) {}
+
+void Account::Initialize() {
+  SHORTFIN_THROW_IF_ERROR(iree_hal_semaphore_create(
+      hal_device(), idle_timepoint_, sem_.for_output()));
+  Reset();
+}
+
+void Account::Reset() {
+  active_tx_type_ = TransactionType::NONE;
+  active_command_buffer_.reset();
+}
+
+void Account::active_deps_extend(iree_hal_semaphore_list_t sem_list) {
+  for (iree_host_size_t i = 0; i < sem_list.count; ++i) {
+    SHORTFIN_THROW_IF_ERROR(iree_hal_fence_insert(
+        active_deps_, sem_list.semaphores[i], sem_list.payload_values[i]));
+  }
+}
+
+// -------------------------------------------------------------------------- //
+// TimelineResource
+// -------------------------------------------------------------------------- //
+
+TimelineResource::TimelineResource(iree_allocator_t host_allocator,
+                                   size_t semaphore_capacity) {
+  SHORTFIN_THROW_IF_ERROR(iree_hal_fence_create(
+      semaphore_capacity, host_allocator, use_barrier_fence_.for_output()));
+}
+
+void TimelineResource::use_barrier_insert(iree_hal_semaphore_t *sem,
+                                          uint64_t timepoint) {
+  SHORTFIN_THROW_IF_ERROR(
+      iree_hal_fence_insert(use_barrier_fence_, sem, timepoint));
+}
+
+// -------------------------------------------------------------------------- //
+// Scheduler
+// -------------------------------------------------------------------------- //
+
+void Scheduler::Initialize(std::span<Device *const> devices) {
+  for (Device *device : devices) {
+    accounts_.emplace_back(device);
+  }
+
+  for (Account &account : accounts_) {
+    auto [it, inserted] = accounts_by_device_id_.emplace(
+        std::make_pair(account.device()->address().device_id(), &account));
+    if (!inserted) {
+      throw std::logic_error("Duplicate device in Scheduler");
+    }
+    account.Initialize();
+    semaphore_count_ += account.semaphore_count();
+  }
+}
+
+Account &Scheduler::GetDefaultAccount(ScopedDevice &device) {
+  auto queue_ordinal = device.affinity().lowest_queue_ordinal();
+  auto device_id =
+      device.raw_device()->address().device_id_for_queue(queue_ordinal);
+  auto it = accounts_by_device_id_.find(device_id);
+  if (it == accounts_by_device_id_.end()) [[unlikely]] {
+    throw std::logic_error(
+        fmt::format("Topology check failed: could not find scheduling account "
+                    "for device id {:x}",
+                    device_id));
+  }
+  return *it->second;
+}
+
+void Scheduler::AppendCommandBuffer(ScopedDevice &device,
+                                    TransactionType tx_type,
+                                    std::function<void(Account &)> callback) {
+  Account &account = GetDefaultAccount(device);
+  logging::info("AppendCommandBuffer({})", static_cast<void *>(&account));
+
+  auto needed_affinity_bits = device.affinity().queue_affinity();
```

Comment:
What is the implication of that here vs at submission? This was one of the things I struggled with and I have a feeling this code will need to be quite switchy based on which transaction mode it is in, since the queue affinity is something that has different usage patterns to it.